### ISSUE TITLE
Bi-directional attention with TTL

### ISSUE THREAD

**Issue by kishorenc** [test-time-training/ttt-lm-jax#9]
The paper states that:

> Instead of handcrafting a self-supervised task from human priors, we take a more end-to-end approach – directly optimizing the self-supervised task for the final goal of next-token prediction.

Does this approach work if I wish to build a BERT-like model that uses bi-directional attention mechanism?

**Comment by karan-dalal** [comment]
Yes. The causality of our method is implicit in our choice of descent mechanism.


<img width="611" alt="Screenshot 2024-08-17 at 5 22 48 PM" src="https://github.com/user-attachments/assets/29efd57c-e66e-45ad-9cf2-ee01c74febaf">

From Fig 7, we take individual gradients on each token, and perform descent sequentially. This prevents any token from using the gradient of future tokens.

For a BERT-style attention mechanism, you could do a variant of "batch gradient descent" instead, where you use all tokens to take 1 gradient step. We did something similar in a CV paper with the method - https://arxiv.org/pdf/2310.13807 (Sec. 5.1)
